src/main/java/at/ac/tuwien/kr/alpha/solver/learning/GroundConflictNoGoodLearner.java

```diff
@@ -110,174 +117,144 @@ public GroundConflictNoGoodLearner(Assignment assignment) {
-				// Only one remaining literals to process, we reached 1UIP.
-				int backjumpingDecisionLevel = computeBackjumpingDecisionLevel(currentResolutionNoGood);
-				if (backjumpingDecisionLevel < 0) {
-					return repeatAnalysisIfNotAssigning(currentResolutionNoGood, resolutionAtoms);
-				}
-				return new ConflictAnalysisResult(currentResolutionNoGood, backjumpingDecisionLevel, false, resolutionAtoms);
-			} else if (firstUIPPriorityQueue.size() < 1) {
-				// This can happen if some NoGood implied a literal at a higher decision level and later the implying literals become (re-)assigned at lower decision levels.
-				// Lowering the decision level may be possible but requires further analysis.
-				// For the moment, just report the learned NoGood.
-				return repeatAnalysisIfNotAssigning(currentResolutionNoGood, resolutionAtoms);
-				//return new ConflictAnalysisResult(currentResolutionNoGood, computeBackjumpingDecisionLevel(currentResolutionNoGood), false, noGoodsResponsible);
-			} else if (currentResolutionNoGood.size() > 32) {
-				// Break if resolved NoGood becomes too large.
-				// Remove all current-dl elements from the resolution NoGood and add the last choice, then backjump like usual.
-				return new ConflictAnalysisResult(null, conflictDecisionLevel, true, resolutionAtoms);	// Flag unsatisfiable abused here.
-				/*if (getAssignmentEntryRespectingLowerMBT(lastGuessedAtom).getDecisionLevel() <= conflictDecisionLevel) {
-					// If lastGuessedAtom is not unassigned after backjump, use repeatAnalysisIfNotAssigning.
-					return repeatAnalysisIfNotAssigning(replaceAllFromConflictDecisionLevelWithGuess(currentResolutionNoGood, conflictDecisionLevel, lastGuessedAtom), noGoodsResponsible, lastGuessedAtom);
+	private String printReasons(Collection<Integer> reasons) {
+		StringBuilder sb = new StringBuilder("{");
+		for (int reasonLiteral : reasons) {
+			sb.append(isPositive(reasonLiteral) ? "+" + atomOf(reasonLiteral) : "-" + atomOf(reasonLiteral));
+			sb.append("@");
+			sb.append(assignment.getWeakDecisionLevel(atomOf(reasonLiteral)));
+			sb.append(", ");
+		}
+		sb.append("}");
+		return sb.toString();
+	}
+
+	private String printReasons(int[] reasons) {
+		return printReasons(IntStream.of(reasons != null ? reasons : new int[0]).boxed().collect(Collectors.toList()));
+	}
+
+	private ConflictAnalysisResult analyzeTrailBased(Antecedent conflictReason) {
+		LOGGER.trace("Analyzing trail based.");
+		if (assignment.getDecisionLevel() == 0) {
+			return ConflictAnalysisResult.UNSAT;
+		}
+		int numLiteralsInConflictLevel = 0;
+		ArrayList<Integer> resolutionLiterals = new ArrayList<>();
+		ArrayList<Integer> resolutionAtoms = new ArrayList<>();
+		int currentDecisionLevel = assignment.getDecisionLevel();
+		HashSet<Integer> seenAtoms = new HashSet<>();		// NOTE: other solvers use a global array for seen atoms, this might be slightly faster (initial tests with local arrays showed no significant improvement).
```

Comment:
```suggestion
		Set<Integer> seenAtoms = new HashSet<>();		// NOTE: other solvers use a global array for seen atoms, this might be slightly faster (initial tests with local arrays showed no significant improvement).
```